### PR TITLE
nodehun: init at 3.0.2

### PR DESCRIPTION
**Pull request by thomasjm**
###### Description of changes

This Node package is used for spellchecking using [Hunspell](https://github.com/hunspell/hunspell), the industry-standard spellchecking package. I think it's useful to have in Nixpkgs because it has native dependencies which are somewhat tricky to build.

It's packaged using the newer `buildNpmPackage` method and the derivation is careful to minimize the closure size.

Any chance I could get you to take a look @natsukium ? This is similar to my previous #237510.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/nodehun/default.nix [hunk 1]
```diff
@@ -0,0 +1,49 @@
+{ buildNpmPackage
+, fetchFromGitHub
+, lib
+, nodePackages
+, nodejs
+, python3
+}:
+
+buildNpmPackage rec {
```
Comment (drupol) [inline pkgs/development/tools/nodehun/default.nix:9]: Can we use the `finalAttrs` pattern with `buildNpmPackage` ?

Let me share a couple of useful links regarding this:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

Comment (thomasjm) [reply pkgs/development/tools/nodehun/default.nix]: The `rec` wasn't needed so I just removed it. Thanks for the review!

**Review by drupol — APPROVED** [review]

**Comment by drupol** [comment]
@thomasjm

2 things:

1. These failures on Darwin arch, are they expected?
2. Could you please squash your commits into one?

**Review by natsukium — COMMENTED** [review]
The darwin build is a bit difficult.
I found that we need `darwin.cctools` for `nativeBuildInputs` and `darwin.apple_sdk.frameworks.CoreServices` for `buildInputs`.
But I still can't build it. It would be better to ping darwin maintainers or mark it as broken.

**Comment by reckenrode** [comment]
@drupol reached out to the Darwin room on the Matrix. I’m taking a look at the Darwin build now.

**Comment by reckenrode** [comment]
@ofborg build hunspell x86_64-linux

**Comment by reckenrode** [comment]
This must be why it’s failing on Darwin but not Linux. It has to build stuff from source on Darwin, and the versions of those things are too old for the nodejs in nixpkgs.

```
npm ERR! node-pre-gyp WARN Using needle for node-pre-gyp https download
npm ERR! node-pre-gyp info check checked for "/private/tmp/nix-shell.LwjqAT/source/node_modules/fsevents/lib/binding/Release/node-v108-darwin-arm64/fse.node" (not found)
npm ERR! node-pre-gyp http GET https://fsevents-binaries.s3-us-west-2.amazonaws.com/v1.2.9/fse-v1.2.9-node-v108-darwin-arm64.tar.gz
npm ERR! node-pre-gyp http 403 https://fsevents-binaries.s3-us-west-2.amazonaws.com/v1.2.9/fse-v1.2.9-node-v108-darwin-arm64.tar.gz
npm ERR! node-pre-gyp WARN Tried to download(403): https://fsevents-binaries.s3-us-west-2.amazonaws.com/v1.2.9/fse-v1.2.9-node-v108-darwin-arm64.tar.gz
npm ERR! node-pre-gyp WARN Pre-built binaries not found for fsevents@1.2.9 and node@18.16.1 (node-v108 ABI, unknown) (falling back to source compile with node-gyp)
npm ERR! node-pre-gyp http 403 status code downloading tarball https://fsevents-binaries.s3-us-west-2.amazonaws.com/v1.2.9/fse-v1.2.9-node-v108-darwin-arm64.tar.gz
```

**Comment by reckenrode** [comment]
I am able to build it on Darwin if I run `npm update nan`, copy the updated `packages-lock.json` over the upstream one, and make sure the the `npmDepsHash` is updated accordingly.

**Comment by drupol** [comment]
I'll set this PR in draft until all the changes are made, please remove it from draft when it's done, so my Github PR filter is able to see it again.

File: pkgs/development/tools/nodehun/default.nix [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    rm -rf build/Release/.deps
+    # Remove a development script to eliminate runtime dependency on node
+    rm node_modules/node-addon-api/tools/conversion.js
+  '';
+
+  doInstallCheck = true;
+  postInstallCheck = ''
+    # Smoke check: require() works
+    export NODE_PATH=$out/lib/node_modules
+    echo 'require("nodehun")' | ${nodejs}/bin/node -
```
[the remark this answers is not in the document]
Comment (thomasjm) [reply pkgs/development/tools/nodehun/default.nix]: Done

**Comment by thomasjm** [comment]
Thanks for figuring that out @natsukium and @reckenrode ! I've added the change to the `buildInputs` and a patch for the `package-lock.json`. It builds for me on Linux, let's see if Darwin works...

File: pkgs/development/tools/nodehun/default.nix [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+  src = fetchFromGitHub {
+    owner = "Wulf";
+    repo = "nodehun";
+    rev = "03c9dcf1fcd965031a68553ccaf6487d1fe87f79";
+    hash = "sha256-MoY95lSIQK1K4aIlMdPm93YxJuez9HYx2zlUhHvDao0=";
+  };
+
+  patches = [
+    ./update-nan.patch
```
Comment (drupol) [inline pkgs/development/tools/nodehun/default.nix:23]: Please add a comment explaining why this patch is required.

**Review by drupol — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/nodehun/default.nix (hunk 3 above)
Comment (thomasjm) [reply pkgs/development/tools/nodehun/default.nix:27]: Done

**Review by drupol — APPROVED** [review]

**Pull request merged**